### PR TITLE
Create Python mini-library + command-line version.

### PR DESCRIPTION
**Pull request by wmwv**
Tested on set of Roman data at NERSC.  Took about 20 minutes on a GPU node to run 5 lightcurves in serial.

```
DATADIR=/pscratch/sd/w/wmwv/RomanDESC
python python/RomanDESCForwardModelLightcurves.py 30328322 20202893 30005877 30300185 41024123441 --datadir ${DATADIR} --verbose
```

### CODE REVIEW

**Comment by wmwv** [comment]
@laldoroty Could you take a look at this PR to add a Python mini-library to do AstroPhot-based forward modeling of SN lightcurves and see
1. Does it run?  What do you need to get it to run?
2. What suggestions do you have for the code itself?
3. Do you have any ideas for expanding testing

Note: It's not an installable package (yet?).  But it's one Python file (still) so the import structure isn't complicated.  Making this a package is a future ticket ( #10 ).

**Comment by laldoroty** [comment]
First issue: conda install -c conda-forge webbpsf did not work (package not found error). Pip install worked.

**Comment by wmwv** [comment]
Sorry you hit that hiccup.  Can you point me to where in the documentation I got this wrong?

**Comment by laldoroty** [comment]
Creating the environment in `conda create --name astrophot -c conda-forge python astropy cudatoolkit h5py ipykernel jupyter matplotlib numpy pandas pyyaml pyarrow scipy requests tqdm webbpsf`.

**Comment by wmwv** [comment]
I _did_ have this wrong in the documentation in previous versions and but thought I had fixed it as of the current SHA `b2f7540`.  I've tried searching and can't figure out where this is still wrong.  Can you flag this for me in the line-by-line review?

**Comment by wmwv** [comment]
Ah, are you reviewing the Notebook?  That's still wrong, but it wasn't meant to be part of this PR.

**Comment by laldoroty** [comment]
Yes, reviewing the notebook--oops. I cloned the main branch. I'll switch to looking at [RomanDESCForwardModelLightcurves.py](https://github.com/Roman-Supernova-PIT/sn-forward-model/blob/issues/2/python/RomanDESCForwardModelLightcurves.py).  Anyway,  here's what I wrote for the notebook after that and before you sent the above comment:

------------------------------------------------

The next error I get is at:
``lightcurve_truth = get_truth_table(truth_files, image_info["visit"], transient_id)``

...but it's just ``FileNotFoundError: [Errno 2] No such file or directory: '../data/RomanDESC/truth/H158/1394/Roman_TDS_index_H158_1394_9.txt'``, so I changed the paths so it matched DCC and then it was fine. 

Same thing at ``targets = ap.image.Target_Image_List(make_target(f, coord=transient_coord) for f in image_files)``.

If I fix that error, at the same line of code, I get ``OSError: Environment variable $WEBBPSF_PATH is not set!``, so I think there should be an instruction in the documentation that says to go here: https://webbpsf.readthedocs.io/en/stable/installation.html and install the data files.

**Comment by laldoroty** [comment]
Ran with `python RomanDESCForwardModelLightcurves.py 20172782 --datadir /cwork/mat90/RomanDESC_sims_2024/RomanTDS`.

Error is: `FileNotFoundError: [Errno 2] No such file or directory: '/cwork/mat90/RomanDESC_sims_2024/RomanTDS/transient_info_table.csv'`

Am I misunderstanding what the datadir argument should be?

EDIT: Add separate `infodir`.

**Comment by laldoroty** [comment]
Had to move test_one_lightcurve.sh up a directory. Ran into this error:
```
visit realized_flux    flux       mag    
----- ------------- ---------- ----------
 1394       12894.0 12882.6435 4.79915235
 8340       12901.0 12882.6435 4.79915235
14088       12834.0 12882.6435 4.79915235
29499       13020.0 12882.6435 4.79915235
36445       13038.0 12882.6435 4.79915235
42193       12787.0 12882.6435 4.79915235
Traceback (most recent call last):
  File "/hpc/group/cosmology/lna18/sn-forward-model/python/RomanDESCForwardModelLightcurves.py", line 1002, in <module>
    parse_and_run()
  File "/hpc/group/cosmology/lna18/sn-forward-model/python/RomanDESCForwardModelLightcurves.py", line 992, in parse_and_run
    run_multiple_transients(
  File "/hpc/group/cosmology/lna18/sn-forward-model/python/RomanDESCForwardModelLightcurves.py", line 583, in run_multiple_transients
    run_one_transient(
  File "/hpc/group/cosmology/lna18/sn-forward-model/python/RomanDESCForwardModelLightcurves.py", line 617, in run_one_transient
    targets = ap.image.Target_Image_List(
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/hpc/group/cosmology/lna18/miniconda3/envs/astrophot/lib/python3.12/site-packages/astrophot/image/target_image.py", line 537, in __init__
    super().__init__(*args, **kwargs)
  File "/hpc/group/cosmology/lna18/miniconda3/envs/astrophot/lib/python3.12/site-packages/astrophot/image/image_object.py", line 484, in __init__
    self.image_list = list(image_list)
                      ^^^^^^^^^^^^^^^^
  File "/hpc/group/cosmology/lna18/sn-forward-model/python/RomanDESCForwardModelLightcurves.py", line 618, in <genexpr>
    make_target(
  File "/hpc/group/cosmology/lna18/sn-forward-model/python/RomanDESCForwardModelLightcurves.py", line 329, in make_target
    hdu = fits.open(image_filepath)
          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/hpc/group/cosmology/lna18/miniconda3/envs/astrophot/lib/python3.12/site-packages/astropy/io/fits/hdu/hdulist.py", line 222, in fitsopen
    return HDUList.fromfile(
           ^^^^^^^^^^^^^^^^^
  File "/hpc/group/cosmology/lna18/miniconda3/envs/astrophot/lib/python3.12/site-packages/astropy/io/fits/hdu/hdulist.py", line 486, in fromfile
    return cls._readfrom(
           ^^^^^^^^^^^^^^
  File "/hpc/group/cosmology/lna18/miniconda3/envs/astrophot/lib/python3.12/site-packages/astropy/io/fits/hdu/hdulist.py", line 1168, in _readfrom
    fileobj = _File(
              ^^^^^^
  File "/hpc/group/cosmology/lna18/miniconda3/envs/astrophot/lib/python3.12/site-packages/astropy/io/fits/file.py", line 218, in __init__
    self._open_filename(fileobj, mode, overwrite)
  File "/hpc/group/cosmology/lna18/miniconda3/envs/astrophot/lib/python3.12/site-packages/astropy/io/fits/file.py", line 651, in _open_filename
    self._file = open(self.name, IO_FITS_MODES[mode])
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: '/cwork/mat90/RomanDESC_sims_2024/RomanTDS/images/H158/1394/Roman_TDS_simple_model_H158_1394_9.fits.gz'

```

because the directory should be `/cwork/mat90/RomanDESC_sims_2024/RomanTDS/images/simple_model/H158...`. So, I can still go with the symlinks, but `images` has a subfolder if you want it to mesh with Troxel's folder structure.

**Comment by laldoroty** [comment]
Suggestion: 
```
from astropy.wcs import FITSFixedWarning
import warnings


# The FITSFixedWarning is consequenceless and it will get thrown every single time you deal with a WCS from the OpenUniverse Roman sims. 
warnings.simplefilter('ignore',category=FITSFixedWarning)
```

Other than that, it runs until
```
PSF Calculation completed.
Calculating jitter using gaussian
Jitter: Convolving with Gaussian with sigma=0.012 arcsec
        resulting image peak drops to 0.973 of its previous value
 Adding extension with image downsampled to detector pixel scale.
 Downsampling to detector pixel scale, by 4
 Downsampling to detector pixel scale, by 4
Astropy WCS not tangent plane coordinate system! May not be compatible with AstroPhot.
Astropy WCS not tangent plane coordinate system! May not be compatible with AstroPhot.
Traceback (most recent call last):
  File "/hpc/group/cosmology/lna18/sn-forward-model/python/RomanDESCForwardModelLightcurves.py", line 1002, in <module>
    parse_and_run()
  File "/hpc/group/cosmology/lna18/sn-forward-model/python/RomanDESCForwardModelLightcurves.py", line 992, in parse_and_run
    run_multiple_transients(
  File "/hpc/group/cosmology/lna18/sn-forward-model/python/RomanDESCForwardModelLightcurves.py", line 583, in run_multiple_transients
    run_one_transient(
  File "/hpc/group/cosmology/lna18/sn-forward-model/python/RomanDESCForwardModelLightcurves.py", line 826, in run_one_transient
    model_host_sn.initialize()
  File "/hpc/group/cosmology/lna18/miniconda3/envs/astrophot/lib/python3.12/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/hpc/group/cosmology/lna18/miniconda3/envs/astrophot/lib/python3.12/site-packages/astrophot/utils/decorators.py", line 21, in wrapped
    warnings.filterwarnings("ignore", category=np.VisibleDeprecationWarning)
                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/hpc/group/cosmology/lna18/miniconda3/envs/astrophot/lib/python3.12/site-packages/numpy/__init__.py", line 410, in __getattr__
    raise AttributeError("module {!r} has no attribute "
AttributeError: module 'numpy' has no attribute 'VisibleDeprecationWarning'
```
I guess astrophot is incompatible with the newest version of numpy (2.0.0). Downgrading to last 1.x.x version, 1.26.4 and retrying.

**Comment by laldoroty** [comment]
It runs!

**Comment by wmwv** [comment]
Yes, `test_one_lightcurve.sh` is meant to be run from `sn-forward-model` like

```
sh tests/test_one_lightcurve.sh
```

That would a useful thing to include in the documentation.

**Review by laldoroty — CHANGES_REQUESTED** [review]
I suggest adding:

```
from astropy.wcs import FITSFixedWarning
import warnings

# The FITSFixedWarning is consequenceless and it will get thrown every single time you deal with a WCS from the OpenUniverse Roman sims. 
warnings.simplefilter('ignore',category=FITSFixedWarning)
```

**Comment by wmwv** [comment]
Context manager added for warnings.  Tested and run.  Warnings successfully suppressed.

Remaining warning messages that do still, _correctly_ appear are
```
Astropy WCS not tangent plane coordinate system! May not be compatible with AstroPhot.
Astropy WCS not tangent plane coordinate system! May not be compatible with AstroPhot.
[W NNPACK.cpp:64] Could not initialize NNPACK! Reason: Unsupported hardware.
```

These are good.  The first two are correct that AstroPhot can't (yet) read TAN-SIP.
And the NNPACK is just saying that we're not GPU enabled.  Although we could consider wrapping this to be more clear in the future, I think that's low priority.

Ready for re-review.

File: tests/test_one_lightcurve.sh [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+DATADIR=${HOME}/data/RomanDESCSims
+
+# In INFODIR should be transient_info_table.csv
+#                      transient_host_info_table.csv
+# We have these in tests/data for a few transients, so we can use those to test
+INFODIR=tests/data/RomanDESC
+DATASET=RomanDESC
+oid=41024123441
+# Run from base directoroy
+python python/RomanDESCForwardModelLightcurves.py \
```
Comment (laldoroty) [inline tests/test_one_lightcurve.sh:15]: This should go up a directory first:
```
python ../python/RomanDESCForwardModelLightcurves.py \
```

Also, typo in "directory" in the comment.

**Comment by laldoroty** [comment]
> Yes, test_one_lightcurve.sh is meant to be run from sn-forward-model like

> `sh tests/test_one_lightcurve.sh`

> That would a useful thing to include in the documentation.

Ah, I meant like, when I run ./test_one_lightcurve.sh from sn-forward-model/tests, it throws the error `python: can't open file '/hpc/group/cosmology/lna18/sn-forward-model/tests/python/RomanDESCForwardModelLightcurves.py': [Errno 2] No such file or directory`. 

So this time, instead of moving the file, I changed the line in the sh file to go up a directory first. See review comment.

Hitting this error now:
```
Traceback (most recent call last):
  File "/hpc/group/cosmology/lna18/sn-forward-model/tests/../python/RomanDESCForwardModelLightcurves.py", line 1011, in <module>
    parse_and_run()
  File "/hpc/group/cosmology/lna18/sn-forward-model/tests/../python/RomanDESCForwardModelLightcurves.py", line 1001, in parse_and_run
    run_multiple_transients(
  File "/hpc/group/cosmology/lna18/sn-forward-model/tests/../python/RomanDESCForwardModelLightcurves.py", line 592, in run_multiple_transients
    run_one_transient(
  File "/hpc/group/cosmology/lna18/sn-forward-model/tests/../python/RomanDESCForwardModelLightcurves.py", line 616, in run_one_transient
    transient_info, transient_host = get_transient_info_and_host(transient_id, infodir)
                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/hpc/group/cosmology/lna18/sn-forward-model/tests/../python/RomanDESCForwardModelLightcurves.py", line 224, in get_transient_info_and_host
    transient_info_table = Table.read(transient_info_file, format="csv")
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/hpc/group/cosmology/lna18/miniconda3/envs/astrophot/lib/python3.12/site-packages/astropy/table/connect.py", line 62, in __call__
    out = self.registry.read(cls, *args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/hpc/group/cosmology/lna18/miniconda3/envs/astrophot/lib/python3.12/site-packages/astropy/io/registry/core.py", line 221, in read
    data = reader(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/hpc/group/cosmology/lna18/miniconda3/envs/astrophot/lib/python3.12/site-packages/astropy/io/ascii/connect.py", line 19, in io_read
    return read(filename, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/hpc/group/cosmology/lna18/miniconda3/envs/astrophot/lib/python3.12/site-packages/astropy/utils/decorators.py", line 603, in wrapper
    return function(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/hpc/group/cosmology/lna18/miniconda3/envs/astrophot/lib/python3.12/site-packages/astropy/utils/decorators.py", line 603, in wrapper
    return function(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/hpc/group/cosmology/lna18/miniconda3/envs/astrophot/lib/python3.12/site-packages/astropy/utils/decorators.py", line 603, in wrapper
    return function(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  [Previous line repeated 2 more times]
  File "/hpc/group/cosmology/lna18/miniconda3/envs/astrophot/lib/python3.12/site-packages/astropy/io/ascii/ui.py", line 428, in read
    dat = fast_reader_rdr.read(table)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/hpc/group/cosmology/lna18/miniconda3/envs/astrophot/lib/python3.12/site-packages/astropy/io/ascii/fastbasic.py", line 131, in read
    self.engine = cparser.CParser(
                  ^^^^^^^^^^^^^^^^
  File "astropy/io/ascii/cparser.pyx", line 242, in astropy.io.ascii.cparser.CParser.__cinit__
  File "astropy/io/ascii/cparser.pyx", line 282, in astropy.io.ascii.cparser.CParser.setup_tokenizer
  File "astropy/io/ascii/cparser.pyx", line 133, in astropy.io.ascii.cparser.FileString.__cinit__
FileNotFoundError: [Errno 2] No such file or directory: 'tests/data/RomanDESC/transient_info_table.csv'

```
It might think this is an absolute path...?

File: tests/test_one_lightcurve.sh (hunk 1 above)
Comment (wmwv) [reply tests/test_one_lightcurve.sh:15]: > This should go up a directory first:

That's what "run from base directory" was meant to say.

The idea is that you would run
```
sh tests/test_one_lightcurve.sh
```
like it says in the README.md

This has come up a couple of times, but I'm not sure how to clearly say the expectation that everything is run from the base directory ('sn-forward-model') not the subdirectories ('python', 'tests', etc.).  Suggestions?

**Comment by wmwv** [comment]
It's not meant to be run as

`./test_one_lightcurve.sh`

it's only meant to be run as

`sh tests/test_one_lightcurve.sh`

How should I document this?

**Comment by wmwv** [comment]
The reasons this makes conceptual sense to me is that it's common to run `pytest tests`, so this is the analogous thing for a shell script.  It's easy to make it work in only one way.  It's annoying and complicated to make it work from anywhere for the shell scripts.

**Comment by laldoroty** [comment]
> It's not meant to be run as
> 
> `./test_one_lightcurve.sh`
> 
> it's only meant to be run as
> 
> `sh tests/test_one_lightcurve.sh`
> 
> How should I document this?

Oh! I understand now. I guess just write something like " Run from base directory with `sh tests/test_one_lightcurve.sh`. It's easier to spot things you are supposed to run verbatim if it's in the code block quotes.

**Comment by laldoroty** [comment]
Still getting error:
```
(base) lna18@dcc-login-05  /hpc/group/cosmology/lna18/sn-forward-model $ sh tests/test_one_lightcurve.sh 
> python: can't open file '/hpc/group/cosmology/lna18/sn-forward-model/../python/RomanDESCForwardModelLightcurves.py': [Errno 2] No such file or directory
```

**Comment by laldoroty** [comment]
Hitting new error, hooray!
```
Traceback (most recent call last):
  File "/hpc/group/cosmology/lna18/sn-forward-model/python/RomanDESCForwardModelLightcurves.py", line 1011, in <module>
    parse_and_run()
  File "/hpc/group/cosmology/lna18/sn-forward-model/python/RomanDESCForwardModelLightcurves.py", line 1001, in parse_and_run
    run_multiple_transients(
  File "/hpc/group/cosmology/lna18/sn-forward-model/python/RomanDESCForwardModelLightcurves.py", line 592, in run_multiple_transients
    run_one_transient(
  File "/hpc/group/cosmology/lna18/sn-forward-model/python/RomanDESCForwardModelLightcurves.py", line 618, in run_one_transient
    lightcurve_truth = get_truth_table(truth_files, image_info["visit"], transient_id)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/hpc/group/cosmology/lna18/sn-forward-model/python/RomanDESCForwardModelLightcurves.py", line 198, in get_truth_table
    this_truth_table = Table.read(tf, format="ascii")
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/hpc/group/cosmology/lna18/miniconda3/envs/astrophot/lib/python3.12/site-packages/astropy/table/connect.py", line 62, in __call__
    out = self.registry.read(cls, *args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/hpc/group/cosmology/lna18/miniconda3/envs/astrophot/lib/python3.12/site-packages/astropy/io/registry/core.py", line 221, in read
    data = reader(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/hpc/group/cosmology/lna18/miniconda3/envs/astrophot/lib/python3.12/site-packages/astropy/io/ascii/connect.py", line 19, in io_read
    return read(filename, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/hpc/group/cosmology/lna18/miniconda3/envs/astrophot/lib/python3.12/site-packages/astropy/utils/decorators.py", line 603, in wrapper
    return function(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/hpc/group/cosmology/lna18/miniconda3/envs/astrophot/lib/python3.12/site-packages/astropy/utils/decorators.py", line 603, in wrapper
    return function(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/hpc/group/cosmology/lna18/miniconda3/envs/astrophot/lib/python3.12/site-packages/astropy/utils/decorators.py", line 603, in wrapper
    return function(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  [Previous line repeated 2 more times]
  File "/hpc/group/cosmology/lna18/miniconda3/envs/astrophot/lib/python3.12/site-packages/astropy/io/ascii/ui.py", line 409, in read
    dat = _guess(table, new_kwargs, format, fast_reader)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/hpc/group/cosmology/lna18/miniconda3/envs/astrophot/lib/python3.12/site-packages/astropy/io/ascii/ui.py", line 616, in _guess
    dat = reader.read(table)
          ^^^^^^^^^^^^^^^^^^
  File "/hpc/group/cosmology/lna18/miniconda3/envs/astrophot/lib/python3.12/site-packages/astropy/io/ascii/core.py", line 1422, in read
    self.lines = self.inputter.get_lines(table, newline=newline)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/hpc/group/cosmology/lna18/miniconda3/envs/astrophot/lib/python3.12/site-packages/astropy/io/ascii/core.py", line 334, in get_lines
    with get_readable_fileobj(table, encoding=self.encoding) as fileobj:
  File "/hpc/group/cosmology/lna18/miniconda3/envs/astrophot/lib/python3.12/contextlib.py", line 137, in __enter__
    return next(self.gen)
           ^^^^^^^^^^^^^^
  File "/hpc/group/cosmology/lna18/miniconda3/envs/astrophot/lib/python3.12/site-packages/astropy/utils/data.py", line 366, in get_readable_fileobj
    fileobj = io.FileIO(name_or_obj, "r")
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: '/hpc/home/lna18/data/RomanDESCSims/truth/H158/1394/Roman_TDS_index_H158_1394_9.txt'

```

So something in the Python script is bumping me back to my home directory instead of `/hpc/group/cosmology/lna18/sn-forward-model/tests/data/RomanDESC/truth`.

**Review by laldoroty — CHANGES_REQUESTED** [review]
Explicitly write "Run from root directory, `sn-forward-model`, with `sh tests/test_one_lightcurve.sh`.

**Comment by wmwv** [comment]
> So something in the Python script is bumping me back to my home directory instead of /hpc/group/cosmology/lna18/sn-forward-model/tests/data/RomanDESC/truth.

Yes, the setting of ${DATADIR} in `test_one_lightcurve.sh` needs to be correct.

It's set to:

```
DATADIR=${HOME}/data/RomanDESCSims
```

I can't make it general in the script.  It needs to be set by the user.

The current text says

```
# Update the variables in the script to match your local system locations for DATADIR.
# In DATADIR should be the subdirectories 'images' and 'truth'.
DATADIR=${HOME}/data/RomanDESCSims
```

Can I make this more clear?

**Review by laldoroty — APPROVED** [review]

**Comment by laldoroty** [comment]
Ah, ok. Actually, I think I would add the instructions for running at the top of the .py file, as well. I missed that.

**Comment by laldoroty** [comment]
Now that i've fixed the path, it runs

**Pull request merged**